src/Streamly/External/ByteString.hs

```diff
@@ -13,45 +14,51 @@ where
 
 import Control.Monad.IO.Class (MonadIO)
 import Data.Word (Word8)
-#if MIN_VERSION_base(4, 10, 0)
-import Foreign.ForeignPtr (plusForeignPtr)
-#else
-import Foreign.ForeignPtr.Compat (plusForeignPtr)
-#endif
-import Foreign.ForeignPtr.Unsafe (unsafeForeignPtrToPtr)
-import GHC.Ptr (minusPtr, plusPtr)
+import GHC.ForeignPtr (ForeignPtr(..))
+import GHC.Ptr (Ptr(..), minusPtr, nullPtr, plusPtr)
 import Streamly.Data.Unfold (Unfold, lmap)
 import Streamly.Data.Fold (Fold)
 
 -- Internal imports
 import Data.ByteString.Internal (ByteString(..))
 import Streamly.Internal.Data.Array.Foreign.Type (Array(..))
+import Streamly.Internal.Data.Array.Foreign.Mut.Type
+    (ArrayContents, arrayToFptrContents, fptrToArrayContents, nilArrayContents)
 
 import qualified Streamly.Data.Array.Foreign as A
 
 import Prelude hiding (read)
 
+-- | Helper function that creates a ForeignPtr
+makeForeignPtr :: ArrayContents -> Ptr a -> ForeignPtr a
+makeForeignPtr contents (Ptr addr#) =
+    ForeignPtr addr# (arrayToFptrContents contents)
+
 -- | Convert a 'ByteString' to an array of 'Word8'. This function unwraps the
 -- 'ByteString' and wraps it with 'Array' constructors and hence the operation
 -- is performed in constant time.
 {-# INLINE toArray #-}
 toArray :: ByteString -> Array Word8
-toArray (PS fp off len) = Array nfp endPtr
+toArray (PS (ForeignPtr addr# _) _ _)
+    | Ptr addr# == nullPtr = Array nilArrayContents nullPtr nullPtr
+toArray (PS (ForeignPtr addr# fpcontents) off len) =
+    Array (fptrToArrayContents fpcontents) startPtr endPtr
   where
-    nfp = fp `plusForeignPtr` off
-    endPtr = unsafeForeignPtrToPtr nfp `plusPtr` len
+    startPtr = Ptr addr# `plusPtr` off
+    endPtr = startPtr `plusPtr` len
 
 -- | Convert an array of 'Word8' to a 'ByteString'. This function unwraps the
 -- 'Array' and wraps it with 'ByteString' constructors and hence the operation
 -- is performed in constant time.
 {-# INLINE fromArray #-}
 fromArray :: Array Word8 -> ByteString
+fromArray arr
+    | arrStart arr == nullPtr = mempty
```

Comment:
Why do we need this? `aLen == 0` below should be enough. We never set arrStart to nullPtr anyway.

Reply:
> We never set `arrStart` to nullPtr anyway.

We do on an empty bytestring